### PR TITLE
Silverstripe 4 compatibility

### PR DESCRIPTION
**Pull request by edwilde**
Update the module to be compatible with Silverstripe 4.

This branch has been used on numerous production sites already, so seems reliable.

### CODE REVIEW

**Comment by thats4shaw** [comment]
@madmatt any chance we could look at getting this merged?

File: README.md [hunk 1]
```diff
@@ -1,36 +1,115 @@
...
+* Encrypting and decrypting data on a field-by-field basis has a performance overhead, which may produce undesirable results in your project.
+* This module uses the `defuse/php-encryption` library under the hood, which prefers strong security over performance. Encrypting lots of fields on a `DataObject` can significantly slow down any operations that read or write large amounts of data (for example `ModelAdmin` views in the CMS that render 50+ records at once can take many seconds of processing power just to decrypt fields). Care should be taken to ensure only the minimal set of data is encrypted, and that this data does not need to be used frequently.
+
+
+## Requirements
+* SilverStripe CMS 4
+* Requires a version of silverstripe-framework that includes [this pull request](https://github.com/silverstripe/silverstripe-framework/pull/10047), or for this PR to be manually patched into your version of framework.
```
Comment (Firesphere) [inline README.md]: PR reference not necessary anymore

File: README.md [hunk 2]
```diff
@@ -1,36 +1,115 @@
 # silverstripe-encrypt-at-rest
 
-This module allows data to be encrypted in the database, but be decrypted when extracted from the database, using a
-secret key (hopefully) known only by the web server.
+This module allows Silverstripe CMS ORM data to be encrypted before being stored in the database, and automatically decrypted before using within your application. To do this, we use a secret key known only by the web server.
 
-*Note:* This does not provide significant protection except in the case of database compromise. It should be used as
```
Comment (Firesphere) [inline README.md:6]: This is a significant caveat and should not be removed

**Review by Firesphere — DISMISSED** [review]
Files are marked as "deleted/created", which makes the PR harder to review than necessary.

**Comment by ssmarco** [comment]
I would love this project to continue forward, let us know if there is anything we can help

File: README.md (hunk 2 above)
Comment (madmatt) [reply README.md:6]: This is okay - it's been expanded into a larger bullet point further down (under the new 'Caveats to understand' section).

File: README.md (hunk 1 above)
Comment (madmatt) [reply README.md]: Good spot, thanks @Firesphere. Agree, the PR is merged so we should just target this at `^4.9` (new extension point was added in 4.9.0).

File: composer.json [hunk 3]
```diff
@@ -1,31 +1,33 @@
...
-
     "support": {
         "issues": "https://github.com/madmatt/silverstripe-encrypt-at-rest/issues",
         "source": "https://github.com/madmatt/silverstripe-encrypt-at-rest"
     },
-
     "require": {
-        "silverstripe/framework": "^3.7",
-        "defuse/php-encryption": "v2.2.1"
+        "silverstripe/framework": "^4.1",
```
Comment (madmatt) [inline composer.json]: This needs to be `^4.9.0` now as that's when the `framework` PR was merged.

**Comment by madmatt** [comment]
@ssmarco @edwilde I've reviewed this one last time and there's just one super minor tweak to make:

1. Please fix up the composer.json min-version for `silverstripe/framework` to be ^4.9.0


I've manually diffed the following files that didn't get marked as a rename and confirmed I'm happy enough with the changes that it doesn't block merging:
1. `code/AtRestCryptoService.php -> src/AtRestCryptoService.php`
2. `code/extensions/EncryptDataObjectsFieldExtension.php -> src/Extension/DecryptDataObjectFieldsExtension.php`
3. `code/fieldtypes/EncryptedDatetime.php` -> `src/FieldType/EncryptedDatetime.php`
4. Removal of `tests/unit/EncryptedDataObjectTest.php` in favour of individual tests for the `AtRestCryptoService`, the extension and field types.

**Comment by madmatt** [comment]
> Files are marked as "deleted/created", which makes the PR harder to review than necessary.

Per my comment above, this one is okay now, I've manually reviewed the files that weren't correctly marked and can confirm there's nothing bad in there, just updates to method signatures and minor tweaks to how the code works primarily.

File: composer.json (hunk 3 above)
Comment (edwilde) [reply composer.json]: Cheers @madmatt, have updated

**Review by madmatt — APPROVED** [review]

**Comment by madmatt** [comment]
Thanks @edwilde, @Firesphere and everyone else who has contributed to this! Merging this now, and I'll comment again when a new major version is released.

**Pull request merged**